### PR TITLE
Add py.typed and remove some type ignore tags

### PR DESCRIPTION
**Pull request by lopez86**
This should ensure that the package works properly with mypy when imported. There are still quite a few type: ignores left for numpy types that should be cleaned up. This also adds a reset_params() function to the abstract language model so that we're not accessing implementation details like parameters directly here.

### CODE REVIEW

File: pyctcdecode/decoder.py [hunk 1]
```diff
@@ -71,6 +84,19 @@
...
+# Generic float type
+if sys.version_info < (3, 8):
+    NpFloat = Any
+else:
+    if sys.version_info < (3, 9) and not TYPE_CHECKING:
+        NpFloat = Any
+    else:
+        NpFloat = np.floating[NBitBase]
+
+FloatVar = TypeVar("FloatVar", bound=NpFloat)
```
Comment (andrew-titus) [inline pyctcdecode/decoder.py:96]: Can we use `np.float_` here? https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float_

**Review by andrew-titus — COMMENTED** [review]
Question about `np.float_`; looks good otherwise

File: pyctcdecode/decoder.py (hunk 1 above)
Comment (lopez86) [reply pyctcdecode/decoder.py:96]: my preference would be not to declare what kind of float the user is using, so float/double/long double should all be acceptable, which I think is the point of np.floating. The NBitBase helps to use the type hierarchy to ensure that information isn't lost in casts (i.e. float32 --> float64 is ok but not the opposite)

Comment (andrew-titus) [reply pyctcdecode/decoder.py:96]: Makes sense -- thanks

**Review by andrew-titus — APPROVED** [review]
LGTM

**Pull request merged**